app/services/emailService.js

```diff
@@ -0,0 +1,125 @@
+const nodemailer = require('nodemailer');
+const config = require('../../config');
+const PortalEmailModel = require('../models/emailModel');
+
+class EmailService {
+  constructor() {
+    this.transporter = nodemailer.createTransport(config.mail);
+  }
+
+  // send alarm mail
+  async sendAlarmMail(subject, message) {
+    let to = config.alarm.mailto;
+    if (to.length <= 0) return;
+    let params = {
+      "async": false,
+      "to": to,
+      "subject": subject,
+      "text": message,
+      "html": `<p>${message}</p>`
+    };
+    await this.sendEmail(params);
+  }
+
+  // send email
+  async sendEmail(params) {
+    let response = {
+      emailStatus : 0,
+      async: params.async,
+      desc: ''
+    };
+    // 1. save information
+    await this._beforeSendEmail(params);
+
+    // 2. send email
+    if (!params.async) {
+      let sendRes = await this.send(params);
+      response.emailStatus = sendRes.emailStatus;
+      response.desc = sendRes.desc;
+    }
+
+    return response;
+  }
+
+  async _beforeSendEmail(params) {
```

Comment:
Bad naming, beforeSendEmail is the description of a period of a specific functionalities that could be easily understand.

Reply:
You haven't solve this, it could be called as 'preprocessEmail'